### PR TITLE
Improve emulated linux aarch64 tests (and make them runnable on kokoro)

### PR DESCRIPTION
**Pull request by jtattermusch**
- simplify the aarch64 linux test (and also make them run faster)
- make it runnable on kokoro

Example test result: https://fusion2.corp.google.com/invocations/b7182a23-8bc7-47d4-be07-342218027c57/log (obtained by temporarily redirecting one of the PR builds to the linux_aarch64.sh script).

### CODE REVIEW

**Comment by jtattermusch** [comment]
@ejona86  please review.

**Review by ejona86 — COMMENTED** [review]
We're currently producing the aarch64 protoc binary from buildscripts/kokoro/linux_artifacts.sh. Do we want to be testing that binary as part of linux_artifact.sh? Or do we want to move the building of that binary to this new linux_aarch64?

**Comment by jtattermusch** [comment]
> We're currently producing the aarch64 protoc binary from buildscripts/kokoro/linux_artifacts.sh. Do we want to be testing that binary as part of linux_artifact.sh? Or do we want to move the building of that binary to this new linux_aarch64?

Good point. 

Let me understand: currently for other architectures than aarch64, you're running the full java test suite as part of linux_artifact.sh? So the "linux_artifacts" jobs is not just for building artifacts, but basically it does all the testing on linux and as a side effect, it also produces the protoc binaries?
I'll try to think of what's the easiest solution for being able to test the aarch64 protoc binaries as part of the arch64 test suite.
Do the maven artifacts we distribute as part of the grpc-java release process need to come from the linux_artifact.sh or can the process be easily adjusted to also consider the outputs of the linux_aarch64 job (in which case it would probably be the easiest to build everything aarch64 related is this job)?

**Comment by jtattermusch** [comment]
Updated the PR to also build the aarch64 maven artifacts (and test them).
New adhoc test results are here: https://source.cloud.google.com/results/invocations/b2368764-b195-4f59-9caa-a591b009d6aa/log

**Comment by jtattermusch** [comment]
@ejona86  PTAL. Do you like the previous version (with aarch64 artifacts built in the linux_artifacts job) or the current version.
I think currently the issue is that the java test suite on aarch64 seems to be flaky (i've experiences some timeouts etc.), so the new linux_aarch64 job is not guaranteed to pass (which could be a problem for releases where you need the release job to pass to collect all the artifacts).

From that perspective, it might make more sense to go this route:
1.  leave aarch64 artifact build in the linux_artifacts jobs for now 
2. merge this PR in its original version (where -PskipCodegen=true tests run)
3. once the linux_aarch64 job is running continously on kokoro, we can more easily look into the flakiness it exhibits and fix it
4. once the linux_aarch64 job is stable, we can use the approach proposed in this PR and move the aarch64 artifact build to the linux_aarch64 job.

Please let me know which approach you want to take.

File: buildscripts/grpc-java-artifacts-aarch64/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM dockcross/manylinux2014-aarch64
```
Comment (ejona86) [inline buildscripts/grpc-java-artifacts-aarch64/Dockerfile:1]: Please no. It seems like this is turtles all the way down and strongly Python-centric.

Can we just add the following to our existing CentOS 7 docker file (`buildscripts/grpc-java-artifacts/Dockerfile`)?
```
yum install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
yum install -y gcc-c++-aarch64-linux-gnu
```

File: buildscripts/run_arm64_tests_in_docker.sh [hunk 2]
```diff
@@ -3,24 +3,50 @@ set -ex
...
+# build aarch64 protoc artifacts via crosscompilation
+# the corresponding codegen tests will be run under and emulator
+# (thanks to the binfmt_misc registration, the emulator will be automatically
+# used for executing aarch64 binaries, so we can execute the codegen tests
+# even though we are in a x86_64 docker container)
+docker run $DOCKER_ARGS --rm=true -v "${grpc_java_dir}":/grpc-java -w /grpc-java \
+  --user "$(id -u):$(id -g)" \
+  -e "JAVA_OPTS=-Duser.home=/grpc-java/.current-user-home -Djava.util.prefs.userRoot=/grpc-java/.current-user-home/.java/.userPrefs" \
+  grpc-java-artifacts-aarch64 \
+  bash -c "LOCAL_MVN_TEMP=$(mktemp -d) SKIP_TESTS=true ARCH=aarch_64 buildscripts/kokoro/unix.sh"
```
Comment (ejona86) [inline buildscripts/run_arm64_tests_in_docker.sh:24]: `LOCAL_MVN_TEMP=$(mktemp -d)` seems like it does nothing, as unix.sh does the same.

Comment (ejona86) [inline buildscripts/run_arm64_tests_in_docker.sh:24]: Use -e for SKIP_TESTS and ARCH. That way this can avoid `bash -c`

File: buildscripts/run_arm64_tests_in_docker.sh [hunk 3]
```diff
@@ -3,24 +3,50 @@ set -ex
...
+
+# build under x64 docker image to save time over building everything under
+# aarch64 emulator. We've already built and tested the protoc binaries
+# so for the rest of the build we will be using "-PskipCodegen=true"
+# avoid further complicating the build.
+docker run $DOCKER_ARGS --rm=true -v "${grpc_java_dir}":/grpc-java -w /grpc-java \
+  --user "$(id -u):$(id -g)" \
+  -e "JAVA_OPTS=-Duser.home=/grpc-java/.current-user-home -Djava.util.prefs.userRoot=/grpc-java/.current-user-home/.java/.userPrefs" \
+  openjdk:11-jdk-slim-buster \
+  bash -c "./gradlew build -x test -PskipAndroid=true -PskipCodegen=true"
```
Comment (ejona86) [inline buildscripts/run_arm64_tests_in_docker.sh:34]: No need for `bash -c`. Remove it.

File: buildscripts/kokoro/linux_aarch64.sh [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+set -veux -o pipefail
+
+if [[ -f /VERSION ]]; then
+  cat /VERSION
+fi
+
+BASE_DIR="$(pwd)"
+
+# Install gRPC and codegen for the Android examples
```
Comment (ejona86) [inline buildscripts/kokoro/linux_aarch64.sh:10]: Delete this comment? Doesn't seem related to anything here.

File: buildscripts/grpc-java-artifacts-aarch64/Dockerfile (hunk 1 above)
Comment (jtattermusch) [reply buildscripts/grpc-java-artifacts-aarch64/Dockerfile:1]: I undestand the concern. the problem though is that after you install `gcc-c++-aarch64-linux-gnu` in centos 7 you still don't have the corresponding libc and libstdc++ for aarch64 and there seems to be no packages for that (the gcc package itself says it's for "kernel compilation only").  So setting up a crosscompiler to work properly in centos7 doesn't seem trivial (or maybe I'm missing something?). Perhaps that's also the reason why in the original version of code  you're actually running `apt-get install -y g++-aarch64-linux-gnu` directly on the ubuntu kokoro worker (and not in the grpc-java-artifacts image).

I could base the image off of debian (where crosscompiler is more readily available), but then the binary compatibility of the resulting binary wouldn't be that good I think.

The reason why I was using `dockcross/manylinux2014-aarch64` is that it's CentOS7 based, it has a working crosscompiler and it's purpose is to produce widely compatible binaries (and we want that).
If you are aware of a good centos7 base image with a crosscompiler, please let me know.

For now I'm keeping `dockcross/manylinux2014-aarch64` since the other solution doesn't work. Please let me know how you want me to handle this.

**Review by jtattermusch — COMMENTED** [review]
Addressed all comments except the usage of `dockcross/manylinux2014-aarch64` - see comment.

**Comment by jtattermusch** [comment]
Ok it seems that some of the tests are flaky and since we're also concerned about the use of `dockcross/manylinux2014-aarch64`, but we don't have a good alternative, here's what I propose:

- I'll split this PR into 2
- the first part will be for running the java tests (the gradle build and gradle test  docker invocations). That way we will have the test running continusly under an emulator and we'll get a better idea of what's flaky.
- the second part will be moving the aarch64 protoc artifact build (and corresponding tests) from linux_artifacts to linux_aarch64.  With this part, we can finish the conversation about how to best crosscompile protoc.

If we do it this way, we will prevent the linux_release job from becoming flaky (flaky job can complicate releases).

**Comment by jtattermusch** [comment]
> Ok it seems that some of the tests are flaky and since we're also concerned about the use of `dockcross/manylinux2014-aarch64`, but we don't have a good alternative, here's what I propose:
> 
> * I'll split this PR into 2
> * the first part will be for running the java tests (the gradle build and gradle test  docker invocations). That way we will have the test running continusly under an emulator and we'll get a better idea of what's flaky.
> * the second part will be moving the aarch64 protoc artifact build (and corresponding tests) from linux_artifacts to linux_aarch64.  With this part, we can finish the conversation about how to best crosscompile protoc.
> 
> If we do it this way, we will prevent the linux_release job from becoming flaky (flaky job can complicate releases).

I created this PR https://github.com/grpc/grpc-java/pull/8113  that doesn't have the protoc artifacts changes. I think we should merge it first.

File: buildscripts/grpc-java-artifacts-aarch64/Dockerfile (hunk 1 above)
Comment (ejona86) [reply buildscripts/grpc-java-artifacts-aarch64/Dockerfile:1]: grpc-java-artifacts Docker image was originally to get CentOS 5 environment (but recently it became CentOS 6). There was no need for that environment with aarch64. As far as I'm concerned, we don't have to run in a docker image and even if we did, debian-based would be fine. Since we statically link the binaries only the glibc version matters. But we don't need to figure that out now because I agree that this is biting off too much right now.

**Comment by jtattermusch** [comment]
Superseded by https://github.com/grpc/grpc-java/pull/8117.
